platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/AdditionalDataBuilderFactory.java

```diff
@@ -18,81 +18,85 @@
-        .put(GeneralDataSpec.class, new DataTypeAndProvider(
+    public AdditionalDataBuilderFactory() {
+        additionalDataProviders.put(GeneralDataSpec.class, new DataTypeAndProvider(
             GeneralData.class,
             new Function<AdditionalData, AdditionalDataBuilder<? extends AdditionalData>>() {
                 @Override
                 public AdditionalDataBuilder<? extends AdditionalData> apply(@Nullable AdditionalData instance) {
                     return DefaultGeneralData.builder((GeneralData) instance);
                 }
-            }))
-        .put(DeprecationDataSpec.class, new DataTypeAndProvider(
+            }));
+        additionalDataProviders.put(DeprecationDataSpec.class, new DataTypeAndProvider(
             DeprecationData.class,
             new Function<AdditionalData, AdditionalDataBuilder<? extends AdditionalData>>() {
                 @Override
                 public AdditionalDataBuilder<? extends AdditionalData> apply(@Nullable AdditionalData instance) {
                     return DefaultDeprecationData.builder((DeprecationData) instance);
                 }
-            }))
-        .put(TypeValidationDataSpec.class, new DataTypeAndProvider(
+            }));
+        additionalDataProviders.put(TypeValidationDataSpec.class, new DataTypeAndProvider(
             TypeValidationData.class,
             new Function<AdditionalData, AdditionalDataBuilder<? extends AdditionalData>>() {
                 @Override
                 public AdditionalDataBuilder<? extends AdditionalData> apply(@Nullable AdditionalData instance) {
                     return DefaultTypeValidationData.builder((TypeValidationData) instance);
                 }
-            }))
-        .put(PropertyTraceDataSpec.class, new DataTypeAndProvider(
+            }));
+        additionalDataProviders.put(PropertyTraceDataSpec.class, new DataTypeAndProvider(
             PropertyTraceData.class,
             new Function<AdditionalData, AdditionalDataBuilder<? extends AdditionalData>>() {
                 @Override
                 public AdditionalDataBuilder<? extends AdditionalData> apply(@Nullable AdditionalData instance) {
                     return DefaultPropertyTraceData.builder((PropertyTraceData) instance);
                 }
-            }))
-        .build();
+            }));
+    }
+
+    /**
+     * Registers a provider for additional data of the given type.
+     *
+     * @param dataType The type of additional data to provide
+     * @param provider The builder function, which will be called to create a builder for the given additional data type
+     */
+    public void registerAdditionalDataProvider(Class<?> dataType, Function<AdditionalData, AdditionalDataBuilder<? extends AdditionalData>> provider) {
+        additionalDataProviders.put(dataType, new DataTypeAndProvider(dataType, provider));
```

Comment:
🤔 Maybe we should add an assertion here to prevent overwriting existing data providers in order to avoid ambiguities.

Reply:
Good idea.